// brainflow/dist/brainflow.types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BrainFlowClassifiers = exports.BrainFlowMetrics = exports.BrainFlowExitCodes = exports.WaveletTypes = exports.NoiseEstimationLevelTypes = exports.WaveletExtensionTypes = exports.ThresholdTypes = exports.WaveletDenoisingTypes = exports.NoiseTypes = exports.DetrendOperations = exports.WindowOperations = exports.AggOperations = exports.FilterTypes = exports.LogLevels = exports.BrainFlowPresets = exports.IpProtocolTypes = exports.BoardIds = exports.BrainFlowError = void 0;
class BrainFlowError extends Error {
    constructor(errorCode, msg) {
        super("Error code is: " + errorCode + " " + msg);
        this.__proto__ = Error;
        this.exitCode = errorCode;
        Object.setPrototypeOf(this, BrainFlowError.prototype);
    }
}
exports.BrainFlowError = BrainFlowError;
var BoardIds;
(function (BoardIds) {
    BoardIds[BoardIds["NO_BOARD"] = -100] = "NO_BOARD";
    BoardIds[BoardIds["PLAYBACK_FILE_BOARD"] = -3] = "PLAYBACK_FILE_BOARD";
    BoardIds[BoardIds["STREAMING_BOARD"] = -2] = "STREAMING_BOARD";
    BoardIds[BoardIds["SYNTHETIC_BOARD"] = -1] = "SYNTHETIC_BOARD";
    BoardIds[BoardIds["CYTON_BOARD"] = 0] = "CYTON_BOARD";
    BoardIds[BoardIds["GANGLION_BOARD"] = 1] = "GANGLION_BOARD";
    BoardIds[BoardIds["CYTON_DAISY_BOARD"] = 2] = "CYTON_DAISY_BOARD";
    BoardIds[BoardIds["GALEA_BOARD"] = 3] = "GALEA_BOARD";
    BoardIds[BoardIds["GANGLION_WIFI_BOARD"] = 4] = "GANGLION_WIFI_BOARD";
    BoardIds[BoardIds["CYTON_WIFI_BOARD"] = 5] = "CYTON_WIFI_BOARD";
    BoardIds[BoardIds["CYTON_DAISY_WIFI_BOARD"] = 6] = "CYTON_DAISY_WIFI_BOARD";
    BoardIds[BoardIds["BRAINBIT_BOARD"] = 7] = "BRAINBIT_BOARD";
    BoardIds[BoardIds["UNICORN_BOARD"] = 8] = "UNICORN_BOARD";
    BoardIds[BoardIds["CALLIBRI_EEG_BOARD"] = 9] = "CALLIBRI_EEG_BOARD";
    BoardIds[BoardIds["CALLIBRI_EMG_BOARD"] = 10] = "CALLIBRI_EMG_BOARD";
    BoardIds[BoardIds["CALLIBRI_ECG_BOARD"] = 11] = "CALLIBRI_ECG_BOARD";
    BoardIds[BoardIds["NOTION_1_BOARD"] = 13] = "NOTION_1_BOARD";
    BoardIds[BoardIds["NOTION_2_BOARD"] = 14] = "NOTION_2_BOARD";
    BoardIds[BoardIds["GFORCE_PRO_BOARD"] = 16] = "GFORCE_PRO_BOARD";
    BoardIds[BoardIds["FREEEEG32_BOARD"] = 17] = "FREEEEG32_BOARD";
    BoardIds[BoardIds["BRAINBIT_BLED_BOARD"] = 18] = "BRAINBIT_BLED_BOARD";
    BoardIds[BoardIds["GFORCE_DUAL_BOARD"] = 19] = "GFORCE_DUAL_BOARD";
    BoardIds[BoardIds["GALEA_SERIAL_BOARD"] = 20] = "GALEA_SERIAL_BOARD";
    BoardIds[BoardIds["MUSE_S_BLED_BOARD"] = 21] = "MUSE_S_BLED_BOARD";
    BoardIds[BoardIds["MUSE_2_BLED_BOARD"] = 22] = "MUSE_2_BLED_BOARD";
    BoardIds[BoardIds["CROWN_BOARD"] = 23] = "CROWN_BOARD";
    BoardIds[BoardIds["ANT_NEURO_EE_410_BOARD"] = 24] = "ANT_NEURO_EE_410_BOARD";
    BoardIds[BoardIds["ANT_NEURO_EE_411_BOARD"] = 25] = "ANT_NEURO_EE_411_BOARD";
    BoardIds[BoardIds["ANT_NEURO_EE_430_BOARD"] = 26] = "ANT_NEURO_EE_430_BOARD";
    BoardIds[BoardIds["ANT_NEURO_EE_211_BOARD"] = 27] = "ANT_NEURO_EE_211_BOARD";
    BoardIds[BoardIds["ANT_NEURO_EE_212_BOARD"] = 28] = "ANT_NEURO_EE_212_BOARD";
    BoardIds[BoardIds["ANT_NEURO_EE_213_BOARD"] = 29] = "ANT_NEURO_EE_213_BOARD";
    BoardIds[BoardIds["ANT_NEURO_EE_214_BOARD"] = 30] = "ANT_NEURO_EE_214_BOARD";
    BoardIds[BoardIds["ANT_NEURO_EE_215_BOARD"] = 31] = "ANT_NEURO_EE_215_BOARD";
    BoardIds[BoardIds["ANT_NEURO_EE_221_BOARD"] = 32] = "ANT_NEURO_EE_221_BOARD";
    BoardIds[BoardIds["ANT_NEURO_EE_222_BOARD"] = 33] = "ANT_NEURO_EE_222_BOARD";
    BoardIds[BoardIds["ANT_NEURO_EE_223_BOARD"] = 34] = "ANT_NEURO_EE_223_BOARD";
    BoardIds[BoardIds["ANT_NEURO_EE_224_BOARD"] = 35] = "ANT_NEURO_EE_224_BOARD";
    BoardIds[BoardIds["ANT_NEURO_EE_225_BOARD"] = 36] = "ANT_NEURO_EE_225_BOARD";
    BoardIds[BoardIds["ENOPHONE_BOARD"] = 37] = "ENOPHONE_BOARD";
    BoardIds[BoardIds["MUSE_2_BOARD"] = 38] = "MUSE_2_BOARD";
    BoardIds[BoardIds["MUSE_S_BOARD"] = 39] = "MUSE_S_BOARD";
    BoardIds[BoardIds["BRAINALIVE_BOARD"] = 40] = "BRAINALIVE_BOARD";
    BoardIds[BoardIds["MUSE_2016_BOARD"] = 41] = "MUSE_2016_BOARD";
    BoardIds[BoardIds["MUSE_2016_BLED_BOARD"] = 42] = "MUSE_2016_BLED_BOARD";
    BoardIds[BoardIds["EXPLORE_4_CHAN_BOARD"] = 44] = "EXPLORE_4_CHAN_BOARD";
    BoardIds[BoardIds["EXPLORE_8_CHAN_BOARD"] = 45] = "EXPLORE_8_CHAN_BOARD";
    BoardIds[BoardIds["GANGLION_NATIVE_BOARD"] = 46] = "GANGLION_NATIVE_BOARD";
    BoardIds[BoardIds["EMOTIBIT_BOARD"] = 47] = "EMOTIBIT_BOARD";
    BoardIds[BoardIds["GALEA_BOARD_V4"] = 48] = "GALEA_BOARD_V4";
    BoardIds[BoardIds["GALEA_SERIAL_BOARD_V4"] = 49] = "GALEA_SERIAL_BOARD_V4";
    BoardIds[BoardIds["NTL_WIFI_BOARD"] = 50] = "NTL_WIFI_BOARD";
    BoardIds[BoardIds["ANT_NEURO_EE_511_BOARD"] = 51] = "ANT_NEURO_EE_511_BOARD";
    BoardIds[BoardIds["EXPLORE_PLUS_8_CHAN_BOARD"] = 54] = "EXPLORE_PLUS_8_CHAN_BOARD";
    BoardIds[BoardIds["EXPLORE_PLUS_32_CHAN_BOARD"] = 55] = "EXPLORE_PLUS_32_CHAN_BOARD";
    BoardIds[BoardIds["PIEEG_BOARD"] = 56] = "PIEEG_BOARD";
    BoardIds[BoardIds["NEUROPAWN_KNIGHT_BOARD"] = 57] = "NEUROPAWN_KNIGHT_BOARD";
})(BoardIds || (exports.BoardIds = BoardIds = {}));
var IpProtocolTypes;
(function (IpProtocolTypes) {
    IpProtocolTypes[IpProtocolTypes["NO_IP_PROTOCOL"] = 0] = "NO_IP_PROTOCOL";
    IpProtocolTypes[IpProtocolTypes["UDP"] = 1] = "UDP";
    IpProtocolTypes[IpProtocolTypes["TCP"] = 2] = "TCP";
})(IpProtocolTypes || (exports.IpProtocolTypes = IpProtocolTypes = {}));
var BrainFlowPresets;
(function (BrainFlowPresets) {
    BrainFlowPresets[BrainFlowPresets["DEFAULT_PRESET"] = 0] = "DEFAULT_PRESET";
    BrainFlowPresets[BrainFlowPresets["AUXILIARY_PRESET"] = 1] = "AUXILIARY_PRESET";
    BrainFlowPresets[BrainFlowPresets["ANCILLARY_PRESET"] = 2] = "ANCILLARY_PRESET";
})(BrainFlowPresets || (exports.BrainFlowPresets = BrainFlowPresets = {}));
var LogLevels;
(function (LogLevels) {
    LogLevels[LogLevels["LEVEL_TRACE"] = 0] = "LEVEL_TRACE";
    LogLevels[LogLevels["LEVEL_DEBUG"] = 1] = "LEVEL_DEBUG";
    LogLevels[LogLevels["LEVEL_INFO"] = 2] = "LEVEL_INFO";
    LogLevels[LogLevels["LEVEL_WARN"] = 3] = "LEVEL_WARN";
    LogLevels[LogLevels["LEVEL_ERROR"] = 4] = "LEVEL_ERROR";
    LogLevels[LogLevels["LEVEL_CRITICAL"] = 5] = "LEVEL_CRITICAL";
    LogLevels[LogLevels["LEVEL_OFF"] = 6] = "LEVEL_OFF";
})(LogLevels || (exports.LogLevels = LogLevels = {}));
var FilterTypes;
(function (FilterTypes) {
    FilterTypes[FilterTypes["BUTTERWORTH"] = 0] = "BUTTERWORTH";
    FilterTypes[FilterTypes["CHEBYSHEV_TYPE_1"] = 1] = "CHEBYSHEV_TYPE_1";
    FilterTypes[FilterTypes["BESSEL"] = 2] = "BESSEL";
    FilterTypes[FilterTypes["BUTTERWORTH_ZERO_PHASE"] = 3] = "BUTTERWORTH_ZERO_PHASE";
    FilterTypes[FilterTypes["CHEBYSHEV_TYPE_1_ZERO_PHASE"] = 4] = "CHEBYSHEV_TYPE_1_ZERO_PHASE";
    FilterTypes[FilterTypes["BESSEL_ZERO_PHASE"] = 5] = "BESSEL_ZERO_PHASE";
})(FilterTypes || (exports.FilterTypes = FilterTypes = {}));
var AggOperations;
(function (AggOperations) {
    AggOperations[AggOperations["MEAN"] = 0] = "MEAN";
    AggOperations[AggOperations["MEDIAN"] = 1] = "MEDIAN";
    AggOperations[AggOperations["EACH"] = 2] = "EACH";
})(AggOperations || (exports.AggOperations = AggOperations = {}));
var WindowOperations;
(function (WindowOperations) {
    WindowOperations[WindowOperations["NO_WINDOW"] = 0] = "NO_WINDOW";
    WindowOperations[WindowOperations["HANNING"] = 1] = "HANNING";
    WindowOperations[WindowOperations["HAMMING"] = 2] = "HAMMING";
    WindowOperations[WindowOperations["BLACKMAN_HARRIS"] = 3] = "BLACKMAN_HARRIS";
})(WindowOperations || (exports.WindowOperations = WindowOperations = {}));
var DetrendOperations;
(function (DetrendOperations) {
    DetrendOperations[DetrendOperations["NO_DETREND"] = 0] = "NO_DETREND";
    DetrendOperations[DetrendOperations["CONSTANT"] = 1] = "CONSTANT";
    DetrendOperations[DetrendOperations["LINEAR"] = 2] = "LINEAR";
})(DetrendOperations || (exports.DetrendOperations = DetrendOperations = {}));
var NoiseTypes;
(function (NoiseTypes) {
    NoiseTypes[NoiseTypes["FIFTY"] = 0] = "FIFTY";
    NoiseTypes[NoiseTypes["SIXTY"] = 1] = "SIXTY";
    NoiseTypes[NoiseTypes["FIFTY_AND_SIXTY"] = 2] = "FIFTY_AND_SIXTY";
})(NoiseTypes || (exports.NoiseTypes = NoiseTypes = {}));
var WaveletDenoisingTypes;
(function (WaveletDenoisingTypes) {
    WaveletDenoisingTypes[WaveletDenoisingTypes["VISUSHRINK"] = 0] = "VISUSHRINK";
    WaveletDenoisingTypes[WaveletDenoisingTypes["SURESHRINK"] = 1] = "SURESHRINK";
})(WaveletDenoisingTypes || (exports.WaveletDenoisingTypes = WaveletDenoisingTypes = {}));
var ThresholdTypes;
(function (ThresholdTypes) {
    ThresholdTypes[ThresholdTypes["SOFT"] = 0] = "SOFT";
    ThresholdTypes[ThresholdTypes["HARD"] = 1] = "HARD";
})(ThresholdTypes || (exports.ThresholdTypes = ThresholdTypes = {}));
var WaveletExtensionTypes;
(function (WaveletExtensionTypes) {
    WaveletExtensionTypes[WaveletExtensionTypes["SYMMETRIC"] = 0] = "SYMMETRIC";
    WaveletExtensionTypes[WaveletExtensionTypes["PERIODIC"] = 1] = "PERIODIC";
})(WaveletExtensionTypes || (exports.WaveletExtensionTypes = WaveletExtensionTypes = {}));
var NoiseEstimationLevelTypes;
(function (NoiseEstimationLevelTypes) {
    NoiseEstimationLevelTypes[NoiseEstimationLevelTypes["FIRST_LEVEL"] = 0] = "FIRST_LEVEL";
    NoiseEstimationLevelTypes[NoiseEstimationLevelTypes["ALL_LEVELS"] = 1] = "ALL_LEVELS";
})(NoiseEstimationLevelTypes || (exports.NoiseEstimationLevelTypes = NoiseEstimationLevelTypes = {}));
var WaveletTypes;
(function (WaveletTypes) {
    WaveletTypes[WaveletTypes["HAAR"] = 0] = "HAAR";
    WaveletTypes[WaveletTypes["DB1"] = 1] = "DB1";
    WaveletTypes[WaveletTypes["DB2"] = 2] = "DB2";
    WaveletTypes[WaveletTypes["DB3"] = 3] = "DB3";
    WaveletTypes[WaveletTypes["DB4"] = 4] = "DB4";
    WaveletTypes[WaveletTypes["DB5"] = 5] = "DB5";
    WaveletTypes[WaveletTypes["DB6"] = 6] = "DB6";
    WaveletTypes[WaveletTypes["DB7"] = 7] = "DB7";
    WaveletTypes[WaveletTypes["DB8"] = 8] = "DB8";
    WaveletTypes[WaveletTypes["DB9"] = 9] = "DB9";
    WaveletTypes[WaveletTypes["DB10"] = 10] = "DB10";
    WaveletTypes[WaveletTypes["DB11"] = 11] = "DB11";
    WaveletTypes[WaveletTypes["DB12"] = 12] = "DB12";
    WaveletTypes[WaveletTypes["DB13"] = 13] = "DB13";
    WaveletTypes[WaveletTypes["DB14"] = 14] = "DB14";
    WaveletTypes[WaveletTypes["DB15"] = 15] = "DB15";
    WaveletTypes[WaveletTypes["BIOR1_1"] = 16] = "BIOR1_1";
    WaveletTypes[WaveletTypes["BIOR1_3"] = 17] = "BIOR1_3";
    WaveletTypes[WaveletTypes["BIOR1_5"] = 18] = "BIOR1_5";
    WaveletTypes[WaveletTypes["BIOR2_2"] = 19] = "BIOR2_2";
    WaveletTypes[WaveletTypes["BIOR2_4"] = 20] = "BIOR2_4";
    WaveletTypes[WaveletTypes["BIOR2_6"] = 21] = "BIOR2_6";
    WaveletTypes[WaveletTypes["BIOR2_8"] = 22] = "BIOR2_8";
    WaveletTypes[WaveletTypes["BIOR3_1"] = 23] = "BIOR3_1";
    WaveletTypes[WaveletTypes["BIOR3_3"] = 24] = "BIOR3_3";
    WaveletTypes[WaveletTypes["BIOR3_5"] = 25] = "BIOR3_5";
    WaveletTypes[WaveletTypes["BIOR3_7"] = 26] = "BIOR3_7";
    WaveletTypes[WaveletTypes["BIOR3_9"] = 27] = "BIOR3_9";
    WaveletTypes[WaveletTypes["BIOR4_4"] = 28] = "BIOR4_4";
    WaveletTypes[WaveletTypes["BIOR5_5"] = 29] = "BIOR5_5";
    WaveletTypes[WaveletTypes["BIOR6_8"] = 30] = "BIOR6_8";
    WaveletTypes[WaveletTypes["COIF1"] = 31] = "COIF1";
    WaveletTypes[WaveletTypes["COIF2"] = 32] = "COIF2";
    WaveletTypes[WaveletTypes["COIF3"] = 33] = "COIF3";
    WaveletTypes[WaveletTypes["COIF4"] = 34] = "COIF4";
    WaveletTypes[WaveletTypes["COIF5"] = 35] = "COIF5";
    WaveletTypes[WaveletTypes["SYM2"] = 36] = "SYM2";
    WaveletTypes[WaveletTypes["SYM3"] = 37] = "SYM3";
    WaveletTypes[WaveletTypes["SYM4"] = 38] = "SYM4";
    WaveletTypes[WaveletTypes["SYM5"] = 39] = "SYM5";
    WaveletTypes[WaveletTypes["SYM6"] = 40] = "SYM6";
    WaveletTypes[WaveletTypes["SYM7"] = 41] = "SYM7";
    WaveletTypes[WaveletTypes["SYM8"] = 42] = "SYM8";
    WaveletTypes[WaveletTypes["SYM9"] = 43] = "SYM9";
    WaveletTypes[WaveletTypes["SYM10"] = 44] = "SYM10";
})(WaveletTypes || (exports.WaveletTypes = WaveletTypes = {}));
var BrainFlowExitCodes;
(function (BrainFlowExitCodes) {
    BrainFlowExitCodes[BrainFlowExitCodes["STATUS_OK"] = 0] = "STATUS_OK";
    BrainFlowExitCodes[BrainFlowExitCodes["PORT_ALREADY_OPEN_ERROR"] = 1] = "PORT_ALREADY_OPEN_ERROR";
    BrainFlowExitCodes[BrainFlowExitCodes["UNABLE_TO_OPEN_PORT_ERROR"] = 2] = "UNABLE_TO_OPEN_PORT_ERROR";
    BrainFlowExitCodes[BrainFlowExitCodes["SER_PORT_ERROR"] = 3] = "SER_PORT_ERROR";
    BrainFlowExitCodes[BrainFlowExitCodes["BOARD_WRITE_ERROR"] = 4] = "BOARD_WRITE_ERROR";
    BrainFlowExitCodes[BrainFlowExitCodes["INCOMMING_MSG_ERROR"] = 5] = "INCOMMING_MSG_ERROR";
    BrainFlowExitCodes[BrainFlowExitCodes["INITIAL_MSG_ERROR"] = 6] = "INITIAL_MSG_ERROR";
    BrainFlowExitCodes[BrainFlowExitCodes["BOARD_NOT_READY_ERROR"] = 7] = "BOARD_NOT_READY_ERROR";
    BrainFlowExitCodes[BrainFlowExitCodes["STREAM_ALREADY_RUN_ERROR"] = 8] = "STREAM_ALREADY_RUN_ERROR";
    BrainFlowExitCodes[BrainFlowExitCodes["INVALID_BUFFER_SIZE_ERROR"] = 9] = "INVALID_BUFFER_SIZE_ERROR";
    BrainFlowExitCodes[BrainFlowExitCodes["STREAM_THREAD_ERROR"] = 10] = "STREAM_THREAD_ERROR";
    BrainFlowExitCodes[BrainFlowExitCodes["STREAM_THREAD_IS_NOT_RUNNING"] = 11] = "STREAM_THREAD_IS_NOT_RUNNING";
    BrainFlowExitCodes[BrainFlowExitCodes["EMPTY_BUFFER_ERROR"] = 12] = "EMPTY_BUFFER_ERROR";
    BrainFlowExitCodes[BrainFlowExitCodes["INVALID_ARGUMENTS_ERROR"] = 13] = "INVALID_ARGUMENTS_ERROR";
    BrainFlowExitCodes[BrainFlowExitCodes["UNSUPPORTED_BOARD_ERROR"] = 14] = "UNSUPPORTED_BOARD_ERROR";
    BrainFlowExitCodes[BrainFlowExitCodes["BOARD_NOT_CREATED_ERROR"] = 15] = "BOARD_NOT_CREATED_ERROR";
    BrainFlowExitCodes[BrainFlowExitCodes["ANOTHER_BOARD_IS_CREATED_ERROR"] = 16] = "ANOTHER_BOARD_IS_CREATED_ERROR";
    BrainFlowExitCodes[BrainFlowExitCodes["GENERAL_ERROR"] = 17] = "GENERAL_ERROR";
    BrainFlowExitCodes[BrainFlowExitCodes["SYNC_TIMEOUT_ERROR"] = 18] = "SYNC_TIMEOUT_ERROR";
    BrainFlowExitCodes[BrainFlowExitCodes["JSON_NOT_FOUND_ERROR"] = 19] = "JSON_NOT_FOUND_ERROR";
    BrainFlowExitCodes[BrainFlowExitCodes["NO_SUCH_DATA_IN_JSON_ERROR"] = 20] = "NO_SUCH_DATA_IN_JSON_ERROR";
    BrainFlowExitCodes[BrainFlowExitCodes["CLASSIFIER_IS_NOT_PREPARED_ERROR"] = 21] = "CLASSIFIER_IS_NOT_PREPARED_ERROR";
    BrainFlowExitCodes[BrainFlowExitCodes["ANOTHER_CLASSIFIER_IS_PREPARED_ERROR"] = 22] = "ANOTHER_CLASSIFIER_IS_PREPARED_ERROR";
    BrainFlowExitCodes[BrainFlowExitCodes["UNSUPPORTED_CLASSIFIER_AND_METRIC_COMBINATION_ERROR"] = 23] = "UNSUPPORTED_CLASSIFIER_AND_METRIC_COMBINATION_ERROR";
})(BrainFlowExitCodes || (exports.BrainFlowExitCodes = BrainFlowExitCodes = {}));
var BrainFlowMetrics;
(function (BrainFlowMetrics) {
    BrainFlowMetrics[BrainFlowMetrics["MINDFULNESS"] = 0] = "MINDFULNESS";
    BrainFlowMetrics[BrainFlowMetrics["RESTFULNESS"] = 1] = "RESTFULNESS";
    BrainFlowMetrics[BrainFlowMetrics["USER_DEFINED"] = 2] = "USER_DEFINED";
})(BrainFlowMetrics || (exports.BrainFlowMetrics = BrainFlowMetrics = {}));
var BrainFlowClassifiers;
(function (BrainFlowClassifiers) {
    BrainFlowClassifiers[BrainFlowClassifiers["DEFAULT_CLASSIFIER"] = 0] = "DEFAULT_CLASSIFIER";
    BrainFlowClassifiers[BrainFlowClassifiers["USER_DEFINED"] = 1] = "USER_DEFINED";
    BrainFlowClassifiers[BrainFlowClassifiers["ONNX_CLASSIFIER"] = 2] = "ONNX_CLASSIFIER";
})(BrainFlowClassifiers || (exports.BrainFlowClassifiers = BrainFlowClassifiers = {}));
